### PR TITLE
Add documentation for the new account key API

### LINKED ISSUE #629: Add documentation for the new Account Key API
This is a sub-task of #371

## Instructions

Add Documentation for the new Account Key API.

### PR DESCRIPTION
**Pull request by SupunS**
Closes #629 

## Description

Add documentation for the new account key API.

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: docs/language/accounts.mdx [hunk 1]
```diff
@@ -49,10 +70,9 @@ Every account can be accessed through two types:
 
       let contracts: AuthAccount.Contracts
 
-      // Key management
+      // Keys
 
-      fun addPublicKey(_ publicKey: [UInt8])
-      fun removePublicKey(_ index: Int)
```
Comment (turbolent) [inline docs/language/accounts.mdx:55]: Maybe leave these functions document, but indicate that they are now deprecated and the new account key API should be used. Maybe just copy https://github.com/onflow/cadence/pull/635/files#diff-922d8e6ca000c087b0eba7a1886b7bccde23a0457bf3412dda16121293a372aeR194 here 👌

File: docs/language/accounts.mdx [hunk 2]
```diff
@@ -96,12 +134,67 @@ and passing the account that should pay for the account creation for the `payer`
...
+struct AccountKey {
+    let keyIndex: Int
+    let publicKey: PublicKey
+    let hashAlgo: HashAlgorithm
+    let weight: UFix64
+    let isRevoked: Bool
+}
+
+struct PublicKey {
+    let publicKey: PublicKey
```
Comment (turbolent) [inline docs/language/accounts.mdx:158]:
```suggestion
    let publicKey: [UInt8]
```

File: docs/language/accounts.mdx [hunk 3]
```diff
@@ -96,12 +134,67 @@ and passing the account that should pay for the account creation for the `payer`
...
+```
+
+## Account Keys
+An account (both `PublicAccount` and `AuthAccount`) have keys associated with it. An account key has the
+following structure.
+```cadence
+struct AccountKey {
+    let keyIndex: Int
+    let publicKey: PublicKey
+    let hashAlgo: HashAlgorithm
```
Comment (turbolent) [inline docs/language/accounts.mdx:152]:
```suggestion
    let hashAlgorithm: HashAlgorithm
```

File: docs/language/accounts.mdx [hunk 4]
```diff
@@ -96,12 +134,67 @@ and passing the account that should pay for the account creation for the `payer`
...
+    let weight: UFix64
+    let isRevoked: Bool
+}
+
+struct PublicKey {
+    let publicKey: PublicKey
+    let signAlgo: SignatureAlgorithm
```
Comment (turbolent) [inline docs/language/accounts.mdx:159]:
```suggestion
    let signatureAlgorithm: SignatureAlgorithm
```

File: docs/language/accounts.mdx [hunk 5]
```diff
@@ -96,12 +134,67 @@ and passing the account that should pay for the account creation for the `payer`
...
+}
+```
+
+A `PublicKey` can be constructed using the raw key and the signing algorithm.
+```cadence
+let publicKey = PublicKey(
+    publicKey: "010203".decodeHex(),
+    signAlgo: SignatureAlgorithm.ECDSA_P256
```
Comment (turbolent) [inline docs/language/accounts.mdx:167]:
```suggestion
    signatureAlgorithm: SignatureAlgorithm.ECDSA_P256
```

File: docs/language/accounts.mdx [hunk 6]
```diff
@@ -96,12 +134,67 @@ and passing the account that should pay for the account creation for the `payer`
...
 For example, to create an account and have the signer of the transaction pay for the account creation,
 and authorize one key to access the account:
 
+```cadence
+transaction(key: PublicKey) {
+    prepare(signer: AuthAccount) {
+        let account = AuthAccount(payer: signer)
+        account.keys.add(
+            publicKey: key,
+            hashAlgo: HashAlgorithm.SHA3_256,
```
Comment (turbolent) [inline docs/language/accounts.mdx:187]:
```suggestion
            hashAlgorithm: HashAlgorithm.SHA3_256,
```

File: docs/language/accounts.mdx [hunk 7]
```diff
@@ -96,12 +134,67 @@ and passing the account that should pay for the account creation for the `payer`
...
+            weight: 10.0
+        )
+    }
+}
+```
+
+<Callout type="info">
+⚠️  Note: Keys can be added also using the `addPublicKey` function. However, this method is currently
+deprecated and is available only for the backward compatibility. The `addPublicKey` method accepts
+the public key encoded together with their signature algorithm, hashing algorithm and weight.
```
Comment (turbolent) [inline docs/language/accounts.mdx:197]: 👌

File: docs/language/accounts.mdx [hunk 8]
```diff
@@ -110,6 +203,44 @@ transaction(key: [UInt8]) {
...
+</Callout>
+
+
+#### Get Account Keys
+Keys that are added to an account can be retrieved using `get()` function, using the index of
+the key. Revoked keys are always returned,  but they have `isRevoked` field set to true.
+Returns `nil` if there is no key available at the given index. Keys can be retrieved from both
+`PublicAccout` and `AuthAccount`.
+```cadence
+transaction() {
```
Comment (turbolent) [inline docs/language/accounts.mdx:215]: Just my OCD sorry: Maybe add some whitespace and linebreaks (after the end of a sentence), here an in the other sections, to make this easier to edit, e.g.:

```suggestion
#### Get Account Keys
Keys that are added to an account can be retrieved using `get()` function, using the index of
the key. 
Revoked keys are always returned,  but they have `isRevoked` field set to true.
Returns `nil` if there is no key available at the given index. Keys can be retrieved from both
`PublicAccout` and `AuthAccount`.

```cadence
transaction() {
```

**Review by turbolent — COMMENTED** [review]
Nice!

**Comment by turbolent** [comment]
@joshuahannan Could you please also have a look? Thank you!

File: docs/language/accounts.mdx [hunk 9]
```diff
@@ -14,10 +14,30 @@ Every account can be accessed through two types:
...
 
+      // Keys
+      let keys: AuthAccount.Keys
+
       // Storage operations
 
       fun getCapability<T>(_ path: PublicPath): Capability<T>
       fun getLinkTarget(_ path: CapabilityPath): Path?
+
+      struct Keys {
```
Comment (joshuahannan) [inline docs/language/accounts.mdx:25]: This struct is also defined in the `AuthAccount` struct below. Should it be in both places?

File: docs/language/accounts.mdx [hunk 10]
```diff
@@ -14,10 +14,30 @@ Every account can be accessed through two types:
       let storageUsed: UInt64
       let storageCapacity: UInt64
 
+      // Keys
+      let keys: AuthAccount.Keys
```
Comment (joshuahannan) [inline docs/language/accounts.mdx:18]: Should this be defined in both `PublicAccount` and `AuthAccount`?

**Review by joshuahannan — APPROVED** [review]
Looks good! I'm just a little confused why the keys struct is in both `PublicAccount` and `AuthAccount`

File: docs/language/accounts.mdx (hunk 9 above)
Comment (SupunS) [reply docs/language/accounts.mdx:25]: Good point! `PublicAccount.Keys` only supports `get` method, whereas `PublicAccount.Keys` supports `add`, `get`, and `revoke`. Hence they had to be implemented as two different structs.

But I have accidentally added the other two methods also, in the documentation for the `PublicAccount.Keys`. Removed it in 98aaf74.
Thanks for catching that!

File: docs/language/accounts.mdx (hunk 10 above)
Comment (SupunS) [reply docs/language/accounts.mdx]: Yeah it is defined in both: (https://github.com/onflow/cadence/pull/635/files#diff-922d8e6ca000c087b0eba7a1886b7bccde23a0457bf3412dda16121293a372aeR63) due to the same reason as above.

File: docs/language/crypto.md [hunk 11]
```diff
@@ -13,25 +55,25 @@ import Crypto
...
+        hashAlgorithm: HashAlgorithm.SHA3_256,
         weight: 0.5
     )
 
-    let publicKeyB = Crypto.PublicKey(
+    let publicKeyB = PublicKey(
         publicKey:
             "df9609ee588dd4a6f7789df8d56f03f545d4516f0c99b200d73b9a3afafc14de5d21a4fc7a2a2015719dc95c9e756cfa44f2a445151aaf42479e7120d83df956".decodeHex(),
-        signatureAlgorithm: Crypto.ECDSA_P256
+        signatureAlgorithm: SignatureAlogirthm.ECDSA_P256
```
Comment (turbolent) [inline docs/language/crypto.md:72]:
```suggestion
        signatureAlgorithm: SignatureAlgorithm.ECDSA_P256
```

File: docs/language/crypto.md [hunk 12]
```diff
@@ -13,25 +55,25 @@ import Crypto
 pub fun test main() {
     let keyList = Crypto.KeyList()
 
-    let publicKeyA = Crypto.PublicKey(
+    let publicKeyA = PublicKey(
         publicKey:
             "db04940e18ec414664ccfd31d5d2d4ece3985acb8cb17a2025b2f1673427267968e52e2bbf3599059649d4b2cce98fdb8a3048e68abf5abe3e710129e90696ca".decodeHex(),
-        signatureAlgorithm: Crypto.ECDSA_P256
+        signatureAlgorithm: SignatureAlogirthm.ECDSA_P256
```
Comment (turbolent) [inline docs/language/crypto.md:61]:
```suggestion
        signatureAlgorithm: SignatureAlgorithm.ECDSA_P256
```

**Review by turbolent — APPROVED** [review]
Looks good!

**Pull request merged**